File: services/platform-api/src/metrics.rs
//! Prometheus-format request metrics for the platform API. The audit flagged that only the realtime
//! gateway exposed metrics (and even those were ad-hoc JSON, not Prometheus) — this closes the gap
//! for the primary API with a dependency-free recorder: per-(method, matched-route, status) request
//! counts and a global latency histogram, rendered as Prometheus text at `/metrics`.
//!
//! Cardinality is bounded by the route table (matched paths, not raw URLs — `/v1/…/{id}/…` collapses
//! to one series), so a `Mutex<HashMap>` on the count map is fine; the histogram uses atomics.

use std::collections::HashMap;
use std::sync::Mutex;
use std::sync::atomic::{AtomicU64, Ordering};

/// Cumulative "less-than-or-equal" latency buckets in milliseconds (a +Inf bucket is implied by
/// the total count). Chosen to bracket a healthy API (sub-50ms) through slow DB/proxy calls.
const LATENCY_BUCKETS_MS: [u64; 8] = [5, 10, 25, 50, 100, 250, 500, 1000];

pub struct Metrics {
    requests: Mutex<HashMap<(String, String, u16), u64>>,
    latency_le: [AtomicU64; 8],
    latency_sum_ms: AtomicU64,
    latency_count: AtomicU64,
}

impl Default for Metrics {
    fn default() -> Self {
        Self::new()
    }
}

impl Metrics {
    pub fn new() -> Self {
        Self {
            requests: Mutex::new(HashMap::new()),
            latency_le: std::array::from_fn(|_| AtomicU64::new(0)),
            latency_sum_ms: AtomicU64::new(0),
            latency_count: AtomicU64::new(0),
        }
    }

    pub fn record(&self, method: &str, matched_path: &str, status: u16, latency_ms: u64) {
        if let Ok(mut m) = self.requests.lock() {
            *m.entry((method.to_owned(), matched_path.to_owned(), status))
                .or_insert(0) += 1;
        }
        for (i, bound) in LATENCY_BUCKETS_MS.iter().enumerate() {
            if latency_ms <= *bound {
                self.latency_le[i].fetch_add(1, Ordering::Relaxed);
            }
        }
        self.latency_sum_ms.fetch_add(latency_ms, Ordering::Relaxed);
        self.latency_count.fetch_add(1, Ordering::Relaxed);
    }

    /// Render Prometheus text-exposition (v0.0.4).
    pub fn render(&self) -> String {
        let mut out = String::new();
        out.push_str("# HELP http_requests_total Total HTTP requests handled.\n");
        out.push_str("# TYPE http_requests_total counter\n");
        if let Ok(m) = self.requests.lock() {
            for ((method, path, status), count) in m.iter() {
                out.push_str(&format!(
                    "http_requests_total{{method=\"{}\",path=\"{}\",status=\"{}\"}} {}\n",
                    escape(method),
                    escape(path),
                    status,
                    count,
                ));
            }
        }

        out.push_str("# HELP http_request_duration_ms Request latency in milliseconds.\n");
        out.push_str("# TYPE http_request_duration_ms histogram\n");
        let count = self.latency_count.load(Ordering::Relaxed);
        for (i, bound) in LATENCY_BUCKETS_MS.iter().enumerate() {
            out.push_str(&format!(
                "http_request_duration_ms_bucket{{le=\"{}\"}} {}\n",
                bound,
                self.latency_le[i].load(Ordering::Relaxed),
            ));
        }
        out.push_str(&format!(
            "http_request_duration_ms_bucket{{le=\"+Inf\"}} {}\n",
            count
        ));
        out.push_str(&format!(
            "http_request_duration_ms_sum {}\n",
            self.latency_sum_ms.load(Ordering::Relaxed)
        ));
        out.push_str(&format!("http_request_duration_ms_count {}\n", count));
        out
    }
}

// Prometheus label values must escape backslash, double-quote, and newline.
fn escape(s: &str) -> String {
    s.replace('\\', "\\\\")
        .replace('"', "\\\"")
        .replace('\n', " ")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn records_and_renders_prometheus_exposition() {
        let m = Metrics::new();
        m.record("GET", "/health", 200, 3);
        m.record("GET", "/health", 200, 12);
        m.record("POST", "/v1/auth/login", 401, 40);

        let out = m.render();
        assert!(
            out.contains("http_requests_total{method=\"GET\",path=\"/health\",status=\"200\"} 2")
        );
        assert!(out.contains(
            "http_requests_total{method=\"POST\",path=\"/v1/auth/login\",status=\"401\"} 1"
        ));
        // 3 requests total; the le="5" bucket holds only the 3ms one.
        assert!(out.contains("http_request_duration_ms_bucket{le=\"5\"} 1"));
        assert!(out.contains("http_request_duration_ms_bucket{le=\"+Inf\"} 3"));
        assert!(out.contains("http_request_duration_ms_count 3"));
        assert!(out.contains("http_request_duration_ms_sum 55"));
    }

    #[test]
    fn escapes_label_values() {
        assert_eq!(escape("a\"b\\c"), "a\\\"b\\\\c");
    }
}

File: services/node-api/lib/metrics.mjs
/**
 * Port of `services/platform-api/src/metrics.rs` — Prometheus text exposition v0.0.4.
 * specs/migration-completion/plan.md §2 (N8)
 *
 * ── One deliberate improvement over the Rust, and why it is safe ────────────────────────────────
 * Rust renders `http_requests_total` by iterating a `HashMap`, whose order is randomized per
 * process. A JS `Map` iterates in INSERTION order, so this renderer is deterministic. That is a
 * divergence, but only in line ORDER, which the Prometheus text format does not constrain — and the
 * A/B compares metric names and bucket bounds rather than bytes precisely because the Rust side
 * cannot be pinned. Deterministic output is strictly easier to diff by hand during a cutover.
 *
 * ── Cardinality ─────────────────────────────────────────────────────────────────────────────────
 * Bounded by the route table: the label is the MATCHED route pattern, never the raw URL, so
 * `/v1/…/{id}/…` collapses to one series regardless of the id. The label uses the AXUM spelling
 * (`{id}`), not Fastify's (`:id`), so a scrape of the Node shell and a scrape of Rust produce
 * comparable series names.
 */

/**
 * Cumulative "less-than-or-equal" latency buckets in milliseconds (a +Inf bucket is implied by the
 * total count). Transcribed from metrics.rs:15 — changing one silently breaks every existing
 * histogram query, so it is pinned by a test rather than left as a literal someone may "tidy".
 */
export const LATENCY_BUCKETS_MS = [5, 10, 25, 50, 100, 250, 500, 1000];

/** Prometheus label values must escape backslash, double-quote, and newline. metrics.rs:95 */
export function escape(s) {
  return s.replaceAll("\\", "\\\\").replaceAll('"', '\\"').replaceAll("\n", " ");
}

export function createMetrics() {
  /** key `${method}\u0000${path}\u0000${status}` → count. A Map keeps insertion order. */
  const requests = new Map();
  const latencyLe = new Array(LATENCY_BUCKETS_MS.length).fill(0);
  let latencySumMs = 0;
  let latencyCount = 0;

  function record(method, matchedPath, status, latencyMs) {
    const key = `${method}\u0000${matchedPath}\u0000${status}`;
    requests.set(key, (requests.get(key) ?? 0) + 1);
    for (let i = 0; i < LATENCY_BUCKETS_MS.length; i += 1) {
      if (latencyMs <= LATENCY_BUCKETS_MS[i]) latencyLe[i] += 1;
    }
    latencySumMs += latencyMs;
    latencyCount += 1;
  }

  function render() {
    let out = "";
    out += "# HELP http_requests_total Total HTTP requests handled.\n";
    out += "# TYPE http_requests_total counter\n";
    for (const [key, count] of requests) {
      const [method, path, status] = key.split("\u0000");
      out += `http_requests_total{method="${escape(method)}",path="${escape(path)}",status="${status}"} ${count}\n`;
    }

    out += "# HELP http_request_duration_ms Request latency in milliseconds.\n";
    out += "# TYPE http_request_duration_ms histogram\n";
    for (let i = 0; i < LATENCY_BUCKETS_MS.length; i += 1) {
      out += `http_request_duration_ms_bucket{le="${LATENCY_BUCKETS_MS[i]}"} ${latencyLe[i]}\n`;
    }
    out += `http_request_duration_ms_bucket{le="+Inf"} ${latencyCount}\n`;
    out += `http_request_duration_ms_sum ${latencySumMs}\n`;
    out += `http_request_duration_ms_count ${latencyCount}\n`;
    return out;
  }

  return { record, render };
}

/**
 * Is the caller allowed to scrape? Port of `metrics_access_allowed` (lib.rs:440).
 *
 * Fail-closed: with a token configured, only that exact token passes. With NO token configured it
 * opens only in explicit dev mode. Anything else 404s — hiding the endpoint's existence rather than
 * advertising it with a 401, because the audit flagged the gateway's /metrics as publicly exposed
 * and this service must not repeat it.
 */
export function metricsAccessAllowed({ metricsToken, metricsDevOpen }, headers) {
  if (metricsToken !== null && metricsToken !== undefined && metricsToken !== "") {
    return headers["x-metrics-token"] === metricsToken;
  }
  return metricsDevOpen === true;
}
